django/db/models/functions/text.py

```diff
@@ -129,14 +128,6 @@ def _paired(self, expressions):
             return ConcatPair(*expressions)
         return ConcatPair(expressions[0], self._paired(expressions[1:]))
 
-    def as_postgresql(self, compiler, connection, **extra_context):
-        sql, params = super().as_sql(compiler, connection, **extra_context)
-
-        # Force text oid instead of unknown for text params
-        params = [Text(p) if isinstance(p, str) else p for p in params]
```

Comment:
@charettes Did you test this with pscopg3? I am not sure how you can get rid of this, The concat tests are still failing…

Reply:
Mhm, let me look over this more closely, I might missed something when rebasing etc…

Reply:
I have rebased this on my psycopg-dev branch and we are still failing in concat:
```
======================================================================
ERROR: test_mixed_char_text (db_functions.text.test_concat.ConcatTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/florian/sources/django.git/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
  File "/home/florian/.local/share/virtualenvs/django/lib/python3.10/site-packages/psycopg/cursor.py", line 725, in execute
    raise ex.with_traceback(None)
psycopg.errors.IndeterminateDatatype: could not determine data type of parameter $1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/florian/sources/django.git/tests/db_functions/text/test_concat.py", line 64, in test_mixed_char_text
    article = Article.objects.annotate(
  File "/home/florian/sources/django.git/django/db/models/query.py", line 650, in get
    num = len(clone)
  File "/home/florian/sources/django.git/django/db/models/query.py", line 380, in __len__
    self._fetch_all()
  File "/home/florian/sources/django.git/django/db/models/query.py", line 1898, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/home/florian/sources/django.git/django/db/models/query.py", line 91, in __iter__
    results = compiler.execute_sql(
  File "/home/florian/sources/django.git/django/db/models/sql/compiler.py", line 1516, in execute_sql
    cursor.execute(sql, params)
  File "/home/florian/sources/django.git/django/db/backends/utils.py", line 103, in execute
    return super().execute(sql, params)
  File "/home/florian/sources/django.git/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
  File "/home/florian/sources/django.git/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/florian/sources/django.git/django/db/backends/utils.py", line 84, in _execute
    with self.db.wrap_database_errors:
  File "/home/florian/sources/django.git/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/florian/sources/django.git/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
  File "/home/florian/.local/share/virtualenvs/django/lib/python3.10/site-packages/psycopg/cursor.py", line 725, in execute
    raise ex.with_traceback(None)
django.db.utils.ProgrammingError: could not determine data type of parameter $1

----------------------------------------------------------------------
(0.001)
INSERT INTO "db_functions_article" ("title",
                                    "summary",
                                    "text",
                                    "written",
                                    "published",
                                    "updated",
                                    "views")
VALUES ('The Title', NULL, '
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
    tempor incididunt ut labore et dolore magna aliqua.', '2022-11-17 03:48:54.692623'::TIMESTAMP, NULL, NULL, 0) RETURNING "db_functions_article"."id";

args=('The Title',
      NONE,
      '\n    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod\n    tempor incididunt ut labore et dolore magna aliqua.',
      datetime.datetime(2022, 11, 17, 3, 48, 54, 692623),
      NONE,
      NONE,
      0);

ALIAS=DEFAULT (0.000)
SELECT "db_functions_article"."id",
       "db_functions_article"."title",
       "db_functions_article"."summary",
       "db_functions_article"."text",
       "db_functions_article"."written",
       "db_functions_article"."published",
       "db_functions_article"."updated",
       "db_functions_article"."views",
       CONCAT("db_functions_article"."title", CONCAT(' - ', "db_functions_article"."text")) AS "title_text"
FROM "db_functions_article"
WHERE "db_functions_article"."title" = 'The Title'
LIMIT 21;

args=(' - ',
      'The Title');

ALIAS=DEFAULT
----------------------------------------------------------------------
Ran 1 test in 0.004s
```

Somehow I have the feeling we'd need to tell the compiler a "compilation context" so we can add the `::text` when passing to a variadic function?

Reply:
@apollo13 sorry for the confusion here, I forgot to push changes I had stashed...

You actually need to keep using `StrDumper` for the `text` OID to be used  for `str` but since there's no more ambiguity for the `JSONField` around encoded values, everything is passed as `Jsonb` instances, it can now properly be used. That means `class Text(str)` can entirely be removed.

I'll push the changes to demonstrate it and has the suite passing for me on ` psycopg[binary]>3`

Reply:
Ah yes, sorry for not seeing that. I just rebased into `psycopg-dev` and will run the tests there.

Reply:
@charettes locally and on Github I am now still failing with:
```
======================================================================
FAIL: test_equal_lowercase (postgres_tests.test_citext.CITextTestCase)
citext removes the need for iexact as the index is case-insensitive.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/florian/sources/django.git/tests/postgres_tests/test_citext.py", line 32, in test_equal_lowercase
    self.assertEqual(
AssertionError: 0 != 1

----------------------------------------------------------------------
(0.001)
SELECT COUNT(*) AS "__count"
FROM "postgres_tests_citestmodel"
WHERE "postgres_tests_citestmodel"."name" = 'john';

args=('john',);
```
I can reproduce this with keepdb if I actually cast `'john'` to text:
```
test_django=# SELECT COUNT(*) AS "__count"
FROM "postgres_tests_citestmodel"
WHERE "postgres_tests_citestmodel"."name" = 'john';      
 __count 
---------
       1
(1 row)

test_django=# SELECT COUNT(*) AS "__count"
FROM "postgres_tests_citestmodel"
WHERE "postgres_tests_citestmodel"."name" = 'john'::text;
 __count 
---------
       0
(1 row)
```

Reply:
https://dba.stackexchange.com/a/181988 :/

Reply:
@apollo13 the CITextField failures should be fixed by the latest commit.

It seems like a sane solution until `CITextField` is entirely removed in Django 5.1.

The rationale here is that it seems less invasive to consider all `str` literal provided to lookups as expressions as `::text` and force the usage of wrappers otherwise (e.g. `CITextField`) than force the usage of wrappers when `::text` is expected.